storage/remote/read_handler.go

```diff
@@ -0,0 +1,271 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"context"
+	"net/http"
+	"sort"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/pkg/gate"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage"
+)
+
+var remoteReadAPIQueries = prometheus.NewGauge(prometheus.GaugeOpts{
+	Namespace: "prometheus",
+	Subsystem: "api",
+	Name:      "remote_read_queries",
+	Help:      "The current number of remote read queries being executed or waiting.",
+})
+
+func init() {
+	prometheus.MustRegister(remoteReadAPIQueries)
+}
+
+type readHandler struct {
+	logger                    log.Logger
+	queryable                 storage.SampleAndChunkQueryable
+	config                    func() config.Config
+	remoteReadSampleLimit     int
+	remoteReadMaxBytesInFrame int
+	remoteReadGate            *gate.Gate
+}
+
+// NewReadHandler creates a http.Handler that accepts remote read requests and
+// writes them to the provided queryable.
+func NewReadHandler(logger log.Logger, queryable storage.SampleAndChunkQueryable, config func() config.Config, remoteReadSampleLimit, remoteReadConcurrencyLimit, remoteReadMaxBytesInFrame int) http.Handler {
+	return &readHandler{
+		logger:                    logger,
+		queryable:                 queryable,
+		config:                    config,
+		remoteReadSampleLimit:     remoteReadSampleLimit,
+		remoteReadGate:            gate.New(remoteReadConcurrencyLimit),
+		remoteReadMaxBytesInFrame: remoteReadMaxBytesInFrame,
+	}
+}
+
+func (h *readHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	if err := h.remoteReadGate.Start(ctx); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	remoteReadAPIQueries.Inc()
+
+	defer h.remoteReadGate.Done()
+	defer remoteReadAPIQueries.Dec()
+
+	req, err := DecodeReadRequest(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	externalLabels := h.config().GlobalConfig.ExternalLabels.Map()
+
+	sortedExternalLabels := make([]prompb.Label, 0, len(externalLabels))
+	for name, value := range externalLabels {
+		sortedExternalLabels = append(sortedExternalLabels, prompb.Label{
+			Name:  name,
+			Value: value,
+		})
+	}
+	sort.Slice(sortedExternalLabels, func(i, j int) bool {
+		return sortedExternalLabels[i].Name < sortedExternalLabels[j].Name
+	})
+
+	responseType, err := NegotiateResponseType(req.AcceptedResponseTypes)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	switch responseType {
+	case prompb.ReadRequest_STREAMED_XOR_CHUNKS:
+		h.remoteReadStreamedXORChunks(ctx, w, req, externalLabels, sortedExternalLabels)
```

Comment:
Would vote for actually returning error and mapping this to `http.Error` here, but not a blocker as we are only moving around.